Integrate RoTTA Method into TTAB Framework

helium777 opened the pull request:
## New Features
- Implemented the Practical Test-Time Adaptation (PTTA) setting as described in the paper, within `scenarios/default_scenarios.py`.
- Integrated the RoTTA (Robust Test-Time Adaptation) method in `model_adaptation/rotta.py`.
- Added functionality for loading the pretrained WideResNet model used in RoTTA, located in `loads/`. 
   - Note: The checkpoint must be manually downloaded from the Google Drive link provided by the RobustBench repository. Additionally, in the pickle file from RobustBench, the key for model weights is labeled 'state_dict' instead of 'model', necessitating a manual conversion for compatibility.
- Provided a minimal working example in `notebooks/rotta_cifar10.ipynb`.

## Considerations
- It's important to note that the implementation of WideResNet in TTAB differs from that in RobustBench, which RoTTA uses. 
   - For instance, in RobustBench, all convolutional layers in WideResNet have a bias, unlike in TTAB.
   - Consequently, the `loads/models` folder now contains two versions of WideresNet: `WideResNet` class in `wideresnet.py`, and `WideResNetRB` class in `wideresnet_robustbench.py`.

## Reproduction Results
I've replicated RoTTA's results on CIFAR-10 to CIFAR-10-C to evaluate its performance under the TTAB framework. It is important to note that I need guidance on calculating the error rate for each domain within the TTAB framework. Therefore, the table includes only partial results:

|              | motion | snow  | fog   | shot  | defocus | contrast | zoom  | brightness | frost | elastic | glass | gaussian | pixelate | jpeg  | impulse | avg.  |
| ------------ | ------ | ----- | ----- | ----- | ------- | -------- | ----- | ---------- | ----- | ------- | ----- | -------- | -------- | ----- | ------- | ----- |
| Reported     | 18.1   | 21.3  | 18.8  | 33.6  | 23.6    | 16.5     | 15.1  | 11.2       | 21.9  | 30.7    | 39.6  | 26.8     | 33.7     | 27.8  | 39.5    | 25.2  |
| Official Repo | 19.72  | 21.59 | 17.66 | 31.58 | 19.37   | 18.50    | 16.07 | 9.38       | 19.97 | 27.52   | 41.36 | 30.08    | 32.06    | 26.18 | 36.47   | 24.50 |
| TTAB         |        |       |       |       |         |          |       |            |       |         |       |          |          |       |         | 31.82 |

## Request for Assistance

I would appreciate guidance on how to properly compute the error rates for each domain within the TTAB framework to complete the above table and verify the implementation's effectiveness.

helium777 committed:
update .gitignore

helium777 committed:
rotta

MarcellusZhao commented on ttab/loads/define_model.py:
```diff
@@ -3,6 +3,7 @@
 
 import timm
 import torch
+from ttab.loads.models.wideresnet_robust_bench import WideResNetRB
```
It's good to test the correctness of implementation using the same model as the paper does, but I think there is no need to integrate this particular model into our benchmark, especially when it has to build another independent model class.

MarcellusZhao commented on ttab/loads/define_model.py:
```diff
@@ -115,7 +116,17 @@ def define_model(config):
 
     # use built-in models and local checkpoints.
-    if "wideresnet" in config.model_name:
+    if "wideresnetrb" in config.model_name:
```
I think we'd better remove this too.

MarcellusZhao commented on ttab/loads/models/wideresnet_robust_bench.py:
```diff
@@ -0,0 +1,97 @@
+        return self.layer(x)
+
+
+class WideResNetRB(nn.Module):
```
Remove this class as well.

MarcellusZhao commented on ttab/model_adaptation/rotta.py:
```diff
@@ -0,0 +1,664 @@
+
+
+# ========== bn_layers ==========
+class MomentumBN(nn.Module):
```
i recommend moving these normalization layers to `model_adaptation/utils.py`.

MarcellusZhao commented on ttab/model_adaptation/rotta.py:
```diff
@@ -0,0 +1,664 @@
+
+
+# ========== custom_transforms ==========
+def get_tta_transforms(cfg, gaussian_std: float = 0.005, soft=False):
```
Why do we need this? it is created for test or something else?

helium777 replied:
It is used to generate the 'augmented view' for the EMA model (the teacher model).

MarcellusZhao replied:
It sounds like a method-specific operation. Probably we can move this into the RoTTA class. What do you think?

MarcellusZhao commented on ttab/model_adaptation/rotta.py:
```diff
@@ -0,0 +1,664 @@
+    return tta_transforms
+
+
+class GaussianNoise(torch.nn.Module):
```
This is irrelevant to this TTA method, right? So please move this to somewhere else or remove it.

helium777 replied:
It is used in the `get_tta_transforms` function, as well as the functions below.

MarcellusZhao replied:
I see. It seems these modules are transferable to other augmentation-based TTA methods. Then it makes sense to move these things to other places.

MarcellusZhao commented on ttab/model_adaptation/rotta.py:
```diff
@@ -0,0 +1,664 @@
+        )
+
+
+class Clip(torch.nn.Module):
```
This is irrelevant as well.

MarcellusZhao commented on ttab/model_adaptation/rotta.py:
```diff
@@ -0,0 +1,664 @@
+        )
+
+
+class ColorJitterPro(ColorJitter):
```
This is irrelevant as well.

MarcellusZhao commented on ttab/model_adaptation/rotta.py:
```diff
@@ -0,0 +1,664 @@
+        return format_string
+
+
+# ========= utils ==========
```
Move irrelevant functions to `utils.py` or make it internal for TTA method class if you have to.

MarcellusZhao commented on ttab/scenarios/default_scenarios.py:
```diff
@@ -36,4 +36,232 @@
             intra_domain_shuffle=True,
         ),
     ),
+    "PTTA_cifar10": Scenario(
```
Too much unnecessary repetition. We'd better develop some new code to support this particular setting.

MarcellusZhao reviewed (commented):
Excellent work! Let's fix comments one by one.

helium777 committed:
Clean up code.

- Remove the implementation of the WideResNet from RobustBench.
- Move some functions in `rotta.py` to either `utils.py` or `RoTTA` class.`

helium777 commented:
## Changes

- The implementation of the WideResNet from RobustBench has been removed.
- Some functions in `rotta.py` have been moved to either `utils.py` or the `RoTTA` class.

## Experiment

I have conducted a quick test for RoTTA using ResNet26 and obtained an error rate of 43.16%. However, I'm unsure if this is a reasonable result.

helium777 committed:
cleanup code
